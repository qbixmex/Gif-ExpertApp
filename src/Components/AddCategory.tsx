import { useState } from "react";

type AddCategoryProps = {
  setCategory: React.Dispatch<React.SetStateAction<string[]>>;
};

const AddCategory = ({ setCategory }: AddCategoryProps) => {
  const [inputValue, setInputValue] = useState<string>("");

  const handleInputChange = (event: React.ChangeEvent<HTMLInputElement>): void => {
    setInputValue(event.target.value);
  };

  const handleSubmit = (event: React.FormEvent<HTMLFormElement>): void => {
    event.preventDefault();

    if (inputValue.trim().length > 2) {
      setCategory( (categories: string[]) => [ inputValue, ...categories ] );
      setInputValue("");
    }
  };

  return (
    <form onSubmit={ handleSubmit }>
      <div className="form-group my-3">
        <input
          type="text"
          className="form-control"
          value={ inputValue }
          onChange={ handleInputChange }
        />
      </div>

      <div className="d-grid">
        <button
          type="submit"
          className="btn btn-primary"
        >
          Add
        </button>
      </div>
    </form>
  );
};

export default AddCategory;
