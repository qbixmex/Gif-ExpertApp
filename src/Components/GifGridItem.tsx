import React from "react";

type GifGridItemProps = {
  id: string;
  title: string;
  url: string;
};

export const GifGridItem = ({ id, title, url }: GifGridItemProps) => {

  return (
    <div className="col-12 col-md-6 col-lg-3">
      <div className="card mb-2 animate__animated animate__fadeIn">
        <img className="img-fluid rounded" id={id} src={url} alt={title} />
        <div className="card-body">
          <h3 id="title" className="card-title black text-center">{title}</h3>
        </div>
      </div>
    </div>
  );
};
