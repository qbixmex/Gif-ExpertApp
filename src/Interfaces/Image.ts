export interface Image {
  id: string;
  title: string;
  images: {
    original: {
      url: string;
    };
  };
}