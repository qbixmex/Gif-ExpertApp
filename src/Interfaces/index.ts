export type { Gif } from './Gif';
export type { Image } from './Image';