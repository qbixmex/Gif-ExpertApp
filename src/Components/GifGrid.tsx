import React from "react";
import { GifGridItem } from './GifGridItem';
import useFetchGif from "../Hooks/useFetchGif";
import "animate.css/animate.min.css";

type GifGridProps = { category: string };

export const GifGrid = ({ category }: GifGridProps) => {

  const { data: images, loading } = useFetchGif( category );

  return (
    <>
    {
      category && loading ?
      <h3 className="text-center display-1 blue animate__animated animate__flash">Loading</h3> :
        <>
          <h2 className="display-2 mb-3 text-center animate__animated animate__fadeIn">{ category }</h2>
          <div className="row">
          {
            images && images.map( image => (
              <GifGridItem
                key={ image.id }
                { ...image }
              />
            ))
          }
          </div>
        </>
    }
    </>
  );
}
