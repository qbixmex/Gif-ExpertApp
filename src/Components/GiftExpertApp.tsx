import { useState } from 'react';
import AddCategory from './AddCategory';
import { GifGrid } from './GifGrid';

type GiftExpertAppProps = {
  defaultCategories?: string[];
};

const GiftExpertApp = ({ defaultCategories = []}: GiftExpertAppProps) => {
  const [categories, setCategories] = useState<string[]>( defaultCategories );

  return (
    <>
      <h1 id="title" className="text-center green">Gif Expert App</h1>

      <AddCategory setCategories={ setCategories } />

      <hr />

      {categories.map(category => (
        <GifGrid
          key={ category }
          category={ category }
        />
      ))}
    </>
  );
};

export default GiftExpertApp;
