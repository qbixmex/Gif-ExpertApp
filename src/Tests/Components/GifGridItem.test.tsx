import { shallow } from "enzyme";
import { GifGridItem } from '../../Components/GifGridItem';
import { Gif } from "../../Interfaces";

const gif: Gif = {
  id: "1234",
  title: "Dragon Ball Super",
  url: "https://giphy.com/api/dragon-ball-super.gif"
};

const wrapper = shallow(
  <GifGridItem
    key={ gif.id }
    { ...gif }
  />
);

// @ts-ignore
describe("Test on GifGridItem", (): void => {
  test("Should load correctly", (): void => {
    expect( wrapper ).toMatchSnapshot();
  });

  test('Should match with title in component', (): void => {
    const title = wrapper.find("#title").text().trim();
    expect( title ).toBe( gif.title );
  });

  test('should to match image properties', (): void => {
    const image = wrapper.find("img");
    expect(image.props().id).toBe(gif.id);
    expect(image.props().alt).toBe(gif.title);
    expect(image.prop("src")).toBe(gif.url);
  });

  test('Should has "animate__animated animate__fadeIn" css classes', (): void => {
    const div = wrapper.find(".card");
    const className = div.prop("className");
    expect(className?.includes("animate__animated animate__fadeIn")).toBe( true );
  }) 
});
