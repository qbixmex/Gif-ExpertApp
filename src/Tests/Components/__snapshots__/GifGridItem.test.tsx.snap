// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test on GifGridItem Should load correctly 1`] = `
<div
  className="col-12 col-md-6 col-lg-3"
>
  <div
    className="card mb-2 animate__animated animate__fadeIn"
  >
    <img
      alt="Dragon Ball Super"
      className="img-fluid rounded"
      id="1234"
      src="https://giphy.com/api/dragon-ball-super.gif"
    />
    <div
      className="card-body"
    >
      <h3
        className="card-title black text-center"
        id="title"
      >
        Dragon Ball Super
      </h3>
    </div>
  </div>
</div>
`;
